File: 2022/01/26/c/removeDuplicates.c
// Program: https://leetcode.com/problems/remove-duplicates-from-sorted-list/
// Author: Andrew Jarombek
// Date: 1/26/2022

#include <stddef.h>
#include <assert.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* deleteDuplicates(struct ListNode* head) {
    if (head == NULL) {
        return head;
    }

    struct ListNode* node = head;

    while (node->next != NULL) {
        if (node->val == node->next->val) {
            node->next = node->next->next;
        } else {
            node = node->next;
        }
    }

    return head;
}

int main() {
    struct ListNode node1 = {3, NULL};
    struct ListNode node2 = {2, &node1};
    struct ListNode node3 = {1, &node2};
    struct ListNode node4 = {1, &node3};

    struct ListNode* result = deleteDuplicates(&node4);
    assert(result->val == 1);
    assert(result->next->val == 2);
    assert(result->next->next->val == 3);
    assert(result->next->next->next == NULL);
}

File: 2023/02/01/removeDuplicates.py
"""
https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
Author: Andrew Jarombek
Date: 2/1/2023
"""

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    @staticmethod
    def delete_duplicates(head: Optional[ListNode]) -> Optional[ListNode]:
        while head is not None and head.next is not None and head.next.val == head.val:
            val = head.val

            while head is not None and head.val == val:
                head = head.next

        result = head

        if result is None:
            return None

        prev = head
        node = head.next

        while node is not None:
            if node.next is not None and node.next.val == node.val:
                val = node.val

                while node is not None and node.val == val:
                    node = node.next

                prev.next = node
            else:
                prev = node
                node = node.next

        return head


if __name__ == '__main__':
    linked_list = ListNode(1, ListNode(1, ListNode(1, ListNode(2, ListNode(3)))))
    result_list = Solution.delete_duplicates(linked_list)
    assert result_list.val == 2
    assert result_list.next.val == 3
    assert result_list.next.next is None
